### PR TITLE
color-mix function

### PR DESCRIPTION
**Pull request by lubegasimon**

### CODE REVIEW

File: packages/reason-css-parser/lib/Parser.re [hunk 1]
```diff
@@ -146,6 +146,19 @@ and color_stop_length = [%value.rec
...
+]
+and polar_color_space = [%value.rec " 'hsl' | 'hwb' | 'lch' | 'oklch' "]
+and rectangular_color_space = [%value.rec
+  " 'srgb' | 'srgb-linear' | 'display-p3' | 'a98-rgb' | 'prophoto-rgb' | 'rec2020' | 'lab' | 'oklab' | 'xyz' | 'xyz-d50' | 'xyz-d65' "
+]
+and color_interpolation_method = [%value.rec
+  " 'in' | [ <rectangular-color-space> | <polar-color-space> <hue-interpolation-method>? ] "
+]
+and function_color_mix = [%value.rec
+  "color-mix(<color-interpolation-method> ',' [ [ <color> && <percentage [0,100]>? ] ]#{2})"
```
Comment (davesnx) [inline packages/reason-css-parser/lib/Parser.re:160]: Where `percentage [0,100]` comes from?

Comment (lubegasimon) [reply packages/reason-css-parser/lib/Parser.re:160]: It's part of the [syntax](https://drafts.csswg.org/css-color-5/#color-mix).

File: packages/reason-css-parser/lib/Parser.re [hunk 2]
```diff
@@ -146,6 +146,19 @@ and color_stop_length = [%value.rec
...
 ]
+and hue_interpolation_method = [%value.rec
+  "[ 'shorter' | 'longer' | 'increasing' | 'decreasing' ] 'hue' "
+]
+and polar_color_space = [%value.rec " 'hsl' | 'hwb' | 'lch' | 'oklch' "]
+and rectangular_color_space = [%value.rec
+  " 'srgb' | 'srgb-linear' | 'display-p3' | 'a98-rgb' | 'prophoto-rgb' | 'rec2020' | 'lab' | 'oklab' | 'xyz' | 'xyz-d50' | 'xyz-d65' "
+]
+and color_interpolation_method = [%value.rec
+  " 'in' | [ <rectangular-color-space> | <polar-color-space> <hue-interpolation-method>? ] "
```
Comment (lubegasimon) [inline packages/reason-css-parser/lib/Parser.re:157]: I think it should be `'in' && ...`

File: packages/reason-css-parser/lib/Parser.re (hunk 1 above)
Comment (davesnx) [reply packages/reason-css-parser/lib/Parser.re:160]: but It's not supported by our parser AFAIK

File: packages/ppx/test/native/Static_test.re [hunk 3]
```diff
@@ -311,6 +311,25 @@ let properties_static_css_tests = [
     [%expr [%css "background-color: red"]],
     [%expr CssJs.backgroundColor(CssJs.red)],
   ),
+  (
+    [%css "background-color: color-mix(in srgb, white)"],
+    [%expr [%css "background-color: color-mix(in srgb, white)"]],
```
Comment (lubegasimon) [inline packages/ppx/test/native/Static_test.re:316]: Still wondering why this errors 
```Error: Property 'background-color' has an invalid value: 'color-mix(in srgb, white)'```

@davesnx, what do you think?

Comment (davesnx) [reply packages/ppx/test/native/Static_test.re:316]: It seems like the parsing logic isn't working. Probably the description is wrong (Parser.re)

Comment (lubegasimon) [reply packages/ppx/test/native/Static_test.re:316]: Indeed, I missed quoting a comma. haha!

Comment (lubegasimon) [reply packages/ppx/test/native/Static_test.re:316]: Thanks

File: packages/css/native/shared/Css_AtomicTypes.ml [hunk 4]
```diff
@@ -865,10 +865,101 @@ module Cursor = struct
...
+  type t = [ `in_ ]
+
+  let toString = function `in_ -> {js|in|js}
+end
+
 module Color = struct
   type t =
     [ `rgb of int * int * int
     | `rgba of int * int * int * [ `num of float | Percentage.t ]
+    | `colorMix of
```
Comment (davesnx) [inline packages/css/native/shared/Css_AtomicTypes.ml:948]: `colorMix2 (x, y)
`colorMix4 (w, x, y z)

File: packages/ppx/src/declarations_to_emotion.re [hunk 5]
```diff
@@ -950,7 +952,129 @@ let render_color = (~loc, value) =>
...
+    let render_size =
+      fun
+      | `Shorter => [%expr `shorter]
+      | `Longer => [%expr `longer]
+      | `Increasing => [%expr `increasing]
+      | `Decreasing => [%expr `decreasing];
+
+    switch (x) {
+    // Should it be a unit 🤔?
+    | (x, ()) => render_size(x)
```
Comment (davesnx) [inline packages/ppx/src/declarations_to_emotion.re:992]: Better to open the tuple on the function argument

File: packages/ppx/test/native/Static_test.re [hunk 6]
```diff
@@ -311,6 +311,118 @@ let properties_static_css_tests = [
...
+  (
+    [%css "background-color: color-mix(in srgb-linear, white, red)"],
+    [%expr [%css "background-color: color-mix(in srgb-linear, white, red)"]],
+    [%expr
+      CssJs.backgroundColor(
+        `colorMix((
+          `method_tup((`in_, `srgbLinear)),
+          `color((CssJs.white, `percent(50.))),
+          `color((CssJs.red, `percent(50.))),
+        )),
```
Comment (davesnx) [inline packages/ppx/test/native/Static_test.re:400]:
```suggestion
        `colorMix2(
          `srgbLinear,
          (CssJs.white, `percent(50.)),
          (CssJs.red, `percent(50.)),
        ),
```

File: packages/ppx/src/declarations_to_emotion.re [hunk 7]
```diff
@@ -950,7 +952,129 @@ let render_color = (~loc, value) =>
...
+      - If both percentage, p1 and p2 are ommited, then p1 = p2 = 50%.
+      - If p1 is omitted, then p1 = 100% - p2.
+      - If p2 is omitted, then p2 = 100% - p1 */
+
+    let render_percentage = (p1, p2) => {
+      switch (p1, p2) {
+      | (Some(p1'), Some(p2')) =>
+        p1' == 0. && p2' == 0.
+          // TODO: Error should be "Invalid function!"
+          ? raise(Unsupported_feature)
```
Comment (davesnx) [inline packages/ppx/src/declarations_to_emotion.re:1022]: raise(Invaid_value("percentatge can be 0"))

File: packages/ppx/src/declarations_to_emotion.re [hunk 8]
```diff
@@ -950,7 +952,129 @@ let render_color = (~loc, value) =>
...
+     https://drafts.csswg.org/css-color-5/#color-mix-percent-norm
+
+      - If p1 + p2 ≠ 100%, then p1' = p1 / (p1 + p2) and p2' = p2 / (p1 + p2),
+          where p1' and p2' are the normalization results.
+      - If p1 = p2 = 0%, the function is invalid
+      - If both percentage, p1 and p2 are ommited, then p1 = p2 = 50%.
+      - If p1 is omitted, then p1 = 100% - p2.
+      - If p2 is omitted, then p2 = 100% - p1 */
+
+    let render_percentage = (p1, p2) => {
```
Comment (davesnx) [inline packages/ppx/src/declarations_to_emotion.re:1017]: let render_percentage = (p1, p2) => {
  switch (p1, p2) {
  | (Some(p1'), Some(p2')) when p1 == 0. and p2' == 0. => raise(Unsupported_feature)
  | (Some(p1'), Some(p2')) when p1' + p2' == 100. => render_percentage(~loc, p1' /. (p1' +. p2'))
  | (Some(p1'), Some(p2')) => render_percentage(~loc, p1')
  | (Some(p1'), None) => render_percentage(~loc, p1')
  | (None, Some(p2')) => render_percentage(~loc, 100. -. p2')
  | (None, None) => render_percentage(~loc, 50.)
  };
};

File: packages/ppx/test/native/Static_test.re [hunk 9]
```diff
@@ -311,6 +311,108 @@ let properties_static_css_tests = [
...
+      )
+    ],
+  ),
+  (
+    [%css "background-color: color-mix(in srgb-linear, white, red)"],
+    [%expr [%css "background-color: color-mix(in srgb-linear, white, red)"]],
+    [%expr
+      CssJs.backgroundColor(
+        `colorMix2((
```
Comment (davesnx) [inline packages/ppx/test/native/Static_test.re:386]: I know I proposed colorMix2/colorMix4, but in reality what makes a bit more sense is to keep colorMix as a single arity but then the first argument of the variant is `in1/`in2/`in3, does that make sense?

Comment (lubegasimon) [reply packages/ppx/test/native/Static_test.re:386]: You mean more or less like this?
```
 `colorMix((
     `in1(`in_, `srgb),
       (CssJs.white, `percent(10.)),
       (CssJs.red, `percent(90.)),
 ))
```

Comment (lubegasimon) [reply packages/ppx/test/native/Static_test.re:386]: if that's the case, `method<>` sounds better as it portrays the name of the argument

Comment (davesnx) [reply packages/ppx/test/native/Static_test.re:386]: More like:

```
 `colorMix((
     `in1(`srgb),
       (CssJs.white, `percent(10.)),
       (CssJs.red, `percent(90.)),
 ))
```

Comment (lubegasimon) [reply packages/ppx/test/native/Static_test.re]: Done!

File: packages/ppx/src/declarations_to_emotion.re [hunk 10]
```diff
@@ -4006,6 +4110,8 @@ let parse_declarations = (~loc: Location.t, property, value) => {
   | Error(_) =>
     switch (render_to_expr(~loc, property, value)) {
     | Ok(value) => Ok(value)
+    | exception (InvalidValue(v)) =>
+      Error(`Invalid_value(value ++ ". " ++ v))
```
Comment (davesnx) [inline packages/ppx/src/declarations_to_emotion.re:4114]: Why the `.` is necessary?

Comment (davesnx) [reply packages/ppx/src/declarations_to_emotion.re:4114]: and I don't see the `\`Invalid_value` error handled anywhere

File: packages/ppx/src/declarations_to_emotion.re [hunk 11]
```diff
@@ -24,6 +24,8 @@ let (let.ok) = Result.bind;
 /* TODO: Add payload on those exceptions */
 exception Unsupported_feature;
 
+exception InvalidValue(string);
```
Comment (davesnx) [inline packages/ppx/src/declarations_to_emotion.re:27]: Let's keep all exceptions with the same casing

s/InvalidValue/Invalid_value

File: packages/ppx/src/declarations_to_emotion.re (hunk 10 above)
Comment (lubegasimon) [reply packages/ppx/src/declarations_to_emotion.re:4114]: The `.` separates the property value from the error message. Which is rendered like this
```
Error: Property 'background-color' has an invalid value: 'color-mix(srgb,
       white 0%, red 0%). Both percentages can not be 0!'
```

Comment (lubegasimon) [reply packages/ppx/src/declarations_to_emotion.re:4114]: hmm! The error message appears inside quotes.

Comment (lubegasimon) [reply packages/ppx/src/declarations_to_emotion.re:4114]: We're calling Invalid_value [here](https://github.com/lubegasimon/styled-ppx/blob/467f391a39743d17f056d1b4a86045b579e99013/packages/ppx/src/declarations_to_emotion.re#L1018).

**Pull request merged**